vllm/v1/core/scheduler.py

```diff
@@ -448,11 +466,36 @@ def update_from_output(
                 new_running.append(request)
                 continue
 
-            request.num_computed_tokens += num_tokens_scheduled
-            # When the request's num_computed_tokens catches up its num_tokens,
-            # the request generates output tokens. Otherwise, we ignore the
-            # sampler output for the request.
-            assert request.num_computed_tokens <= request.num_tokens
+            req_index = model_runner_output.req_id_to_index[req_id]
+            generated_token_ids = sampled_token_ids[req_index]
+            if not use_spec_decode:
```

Comment:
Do you mean this?
```suggestion
            if req_id not in scheduler_output.scheduled_spec_decode_tokens:
```

Reply:
Ah yes. I find `use_spec_decode` a bit confusing. I personally think it's clearer to have it per request.